packages/opentelemetry-scope-async-hooks/src/AsyncHooksScopeManager.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ScopeManager } from '@opentelemetry/scope-base';
+import * as asyncHooks from 'async_hooks';
+import { EventEmitter } from 'events';
+import * as shimmer from 'shimmer';
+
+const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> = [
+  'addListener',
+  'on',
+  'once',
+  'prependListener',
+  'prependOnceListener',
+];
+
+export class AsyncHooksScopeManager implements ScopeManager {
+  private asyncHook: asyncHooks.AsyncHook;
+  private scopes: { [uid: number]: unknown } = {};
+  private current: unknown = null;
+
+  constructor() {
+    this.asyncHook = asyncHooks.createHook({
+      init: this.init.bind(this),
+      before: this.before.bind(this),
+      destroy: this.destroy.bind(this),
+    });
+  }
+
+  active(): unknown {
+    return this.current;
+  }
+
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
+    scope: unknown,
+    fn: T
+  ): ReturnType<T> {
+    const oldCurrent = this.current;
+    this.current = scope;
+    const res = fn();
+    this.current = oldCurrent;
+    return res;
+  }
+
+  bind<T>(target: T, scope?: unknown): T {
+    // if no specific scope to propagate is give, we use the current one
+    if (!scope) {
+      scope = this.current;
+    }
+    if (target instanceof EventEmitter) {
+      return this.bindEventEmitter(target, scope);
+    } else if (typeof target === 'function') {
+      return this.bindFunction(target, scope);
+    }
+    return target;
+  }
+
+  enable(): this {
+    this.asyncHook.enable();
+    return this;
+  }
+
+  disable(): this {
+    this.asyncHook.disable();
+    this.scopes = {};
+    this.current = null;
+    return this;
+  }
+
+  private bindFunction<T extends Function>(target: T, scope?: unknown): T {
+    const manager = this;
+    const boundContext = scope;
+    const contextWrapper = function(this: {}) {
```

Comment:
Most of this is the same as `with()`. Can `with()` be called instead?

Reply:
We can't use `with()` since it would change the `this` of the called method, we need to call the function with the same context as the wrapped method.

Reply:
This can still be done by mixing a regular function with an arrow function. For example:
```js
const contextWrapper = function () {
  return manager.with(scope, () => target.apply(this, arguments))
}
```